fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala

```diff
@@ -38,26 +38,44 @@ object RedisConnection {
     }
 
   private[fs2] final object impl {
+
     def send[F[_]: Applicative](sink: Sink[F, Byte])(implicit log: LogWriter[F]): Sink[F, RESP] =
       _.evalMap(resp => log.debug(s"sending $resp") *> resp.pure)
         .through(streamEncoder.encode)
         .flatMap(bits => Stream.segment(Segment.array(bits.toByteArray)))
         .to(sink)
 
     def receive[F[_]: Effect](implicit log: LogWriter[F]): Pipe[F, Byte, RESP] = {
-      def toBitVector: Pipe[F, Byte, BitVector] = {
-        def go(stream: Stream[F, Byte]): Pull[F, BitVector, Unit] = stream.pull.unconsChunk.flatMap {
-          case Some((chunk, rest)) => Pull.output1(BitVector.view(chunk.toByteBuffer)) >> go(rest)
-          case _                   => Pull.done
-        }
-
-        stream =>
-          go(stream).stream
+
+      def toBitVector: Pipe[F, Byte, NonEmptyFrame] = {
+
+        def go(stream: Stream[F, Byte], previous: Frame): Pull[F, NonEmptyFrame, Unit] =
+          stream.pull.unconsChunk flatMap {
+
+            case Some((chunk, rest)) =>
+              previous.append(chunk) match {
+                case Left(ex) => Pull.raiseError(ex)
+                case Right(f) => f match {
+                  case frame @ (Complete(_) | Decoded(_)) => Pull.output1(frame) >> go(rest, Empty)
+                  case frame: Incomplete                  => go(rest, frame)
+                }
+              }
+
+            case _ => Pull.done
+          }
+
+        stream => go(stream, Empty).stream
       }
 
-      _.through(toBitVector)
-        .flatMap(streamDecoder.decode(_))
-        .evalMap(resp => log.debug(s"receiving $resp") *> resp.pure)
+      _.through(toBitVector) flatMap {
+        case Complete(v)      => streamDecoder.decode(v)
+        case Decoded(s)       => Stream(s)
+        case Incomplete(v, _) => Stream.raiseError(
+          new Exception(s"Trying to decode an incomplete frame. Content: ${v.toByteArray.map(_.toChar).mkString}")
```

Comment:
at some stage we'd probably want to be more precise on the error types wdyt?

Reply:
Yes, just I wouldn't know what to add there. This actually should never happen considering the below

```scala
case Right(f) => f match {
  case frame @ (Complete(_) | Decoded(_)) => Pull.output1(frame) >> go(rest, Empty)
  case frame: Incomplete                                => go(rest, frame)
}
```
I will think to that.

Reply:
is that so though? could it manifest if the response stream is corrupted?
if we don't believe it will occur and/or we should _not_ cater for that then what about adding another marker trait `CompleteFrame` and mixing it into `Complete` and `Decoded`? then you could change the `Pull`'s return type to `Pull[F, CompleteFrame, Unit]` right?

Reply:
Actually you're right, `NonEmptyFrame` is too wide for the `Pull` output. If the data are corrupted they will manifest in a problem in the decoder, the frame should never produce out a vector with less bits than the ones declared in the header.